pkg/monitor/monitor/util.go

```diff
@@ -0,0 +1,679 @@
+			Namespace:       monitor.Namespace,
+			Labels:          monitorLabel,
+			OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
+		},
+		Rules: []rbac.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get", "list", "watch"},
+			},
+			{
+				NonResourceURLs: []string{"/metrics"},
+				Verbs:           []string{"get"},
+			},
+		},
+	}
+}
+
+func getMonitorClusterRoleBinding(sa *core.ServiceAccount, cr *rbac.ClusterRole, monitor *v1alpha1.TidbMonitor) *rbac.ClusterRoleBinding {
+	monitorLabel := label.New().Instance(monitor.Name).Monitor().Labels()
+	return &rbac.ClusterRoleBinding{
+		ObjectMeta: meta.ObjectMeta{
+			Name:            getMonitorObjectName(monitor),
+			Namespace:       monitor.Namespace,
+			Labels:          monitorLabel,
+			OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      sa.Name,
+				Namespace: sa.Namespace,
+				APIGroup:  "",
+			},
+		},
+		RoleRef: rbac.RoleRef{
+			Kind:     "ClusterRole",
+			Name:     cr.Name,
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+}
+
+func getMonitorDeployment(sa *core.ServiceAccount, config *core.ConfigMap, secret *core.Secret, monitor *v1alpha1.TidbMonitor, tc *v1alpha1.TidbCluster) *apps.Deployment {
+	deployment := getMonitorDeploymentSkeleton(sa, monitor)
+	initContainer := getMonitorInitContainer(monitor, tc)
+	deployment.Spec.Template.Spec.InitContainers = append(deployment.Spec.Template.Spec.InitContainers, initContainer)
+	prometheusContainer := getMonitorPrometheusContainer(monitor, tc)
+	reloaderContainer := getMonitorReloaderContainer(monitor, tc)
+	deployment.Spec.Template.Spec.Containers = append(deployment.Spec.Template.Spec.Containers, prometheusContainer, reloaderContainer)
+	if monitor.Spec.Grafana != nil {
+		grafanaContainer := getMonitorGrafanaContainer(secret, monitor, tc)
+		deployment.Spec.Template.Spec.Containers = append(deployment.Spec.Template.Spec.Containers, grafanaContainer)
+	}
+	volumes := getMonitorVolumes(config, monitor, tc)
+	deployment.Spec.Template.Spec.Volumes = volumes
+	return deployment
+}
+
+func getMonitorDeploymentSkeleton(sa *core.ServiceAccount, monitor *v1alpha1.TidbMonitor) *apps.Deployment {
+	monitorLabel := label.New().Instance(monitor.Name).Monitor().Labels()
+	replicas := int32(1)
+
+	deployment := &apps.Deployment{
+		ObjectMeta: meta.ObjectMeta{
+			Name:            getMonitorObjectName(monitor),
+			Namespace:       monitor.Namespace,
+			Labels:          monitorLabel,
+			OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
+			Annotations:     monitor.Spec.Annotations,
+		},
+		Spec: apps.DeploymentSpec{
+			Replicas: &replicas,
+			Strategy: apps.DeploymentStrategy{
+				Type:          apps.RecreateDeploymentStrategyType,
+				RollingUpdate: nil,
```

Comment:
I think the BP is to omit unnecessary fields: https://kubernetes.io/docs/concepts/configuration/overview/#general-configuration-tips

Reply:
updated.